File: clockhands.cpp
#include <iostream>
#include <cmath>

/*
* Exercise to calculate the angle between the hands of the clock at a given time specified by integer hour and minute values.
* This version assumes a standard 12 hour clock with 60 minute hours.
* It assumes the hour hand moves smoothly between the hours.
* Calculation is done twice: 
* method 1 uses the scalar product of the vectors representing the hands. It is more resource intensive (trigonometric functions)
* but requires no further logic to get the angle in the right range, so useful for testing.
* method 2 uses the hand angle wrt hour=12 directly to get the difference between the angles. This requires noting that the
* angle may be negative and/or reentrant and correct this.
* Both answers are printed.
* The program loops over successive inputs of hour and minute from the console until terminated by hour<0.
*/

// Define PI (<cmath> does not appear to have M_PI in the present project setup)
const double twopi = 3.14159265358979323846 * 2;

// Define hours in the clock and minutes in the hour
// At some stage this could be read from a configuration file
const double HOURS_PER_CLOCK = 12.0;
const double MINS_PER_HOUR = 60;

// Define a 2D vector (x,y) and a dot product. The norm() method is provided but not used.
// It is used to calculate the angle between 2 vectors from a DOT b = |a||b|cos theta under one method considered here
class Vec2d
{
	double x, y;
public:
	Vec2d(double _x, double _y) : x(_x), y(_y) {}
	double operator*(const Vec2d& v) { return x * v.x + y * v.y; } // returns dot product
	double norm() { return std::sqrt((*this) * (*this)); }
};

// Class for any hand with a position and an angle
// HourHand and MinuteHand specialise this
class Hand
{
protected:
	// value for time
	double time;
	// value of angle of real 2d vector wrt hour=12
	double angle;
public:
	Hand(double _val) : time(_val), angle(-666.0) {} // angle calc'd in subclasses, initialised to bad value
	Vec2d vector_from_angle() const { return Vec2d(std::cos(angle), std::sin(angle)); }
	double get_angle() const { return angle; }
};

class HourHand : public Hand
{
public:
	HourHand(int _hval, int _mval) : Hand(_hval + _mval / MINS_PER_HOUR) { angle = to_angle(); }
private:
	double to_angle() const { return time * twopi / HOURS_PER_CLOCK; }
};

class MinuteHand : public Hand
{
public:
	MinuteHand(int _val) : Hand(_val) { angle = to_angle(); }
private:
	double to_angle() const { return time * twopi / MINS_PER_HOUR; }
};


int main()
{
	while (true)
	{
		std::cout << "Hello World!\n";
		std::cout << "Input hour value (int, <0 to exit): ";
		int hour;
		std::cin >> hour;
		if (hour < 0) break;
		std::cout << "Input minute value (int): ";
		int min;
		std::cin >> min;
		const HourHand h(hour, min);
		const MinuteHand m(min);
		const double cosangle = h.vector_from_angle() * m.vector_from_angle();
		double angle = std::acos(cosangle) * 360.0 / twopi;
		std::cout << "Angle between hands = " << angle << std::endl;
		const double hangle = h.get_angle();
		const double mangle = m.get_angle();
		angle = std::fabs(hangle - mangle);
		if (angle * 2.0 > twopi) angle = twopi - angle;
		angle *= 360.0 / twopi;
		std::cout << "Angle between hands = " << angle << std::endl;
	}
}

File: clockhands.py
#
# Interview exercise code to calculate the angle between the hands of a clock
# Two calculation methods are compared, using cos of angle from dot product
# and direct difference from angles.
# First method uses more compute (trig functions) but gives answer without need
# to differentiate negative and reentrant angles
#

import math

twopi = 3.141592653589793238463*2.0 #math.pi*2
# explicit literal and pi from math give same answers, just below exact values (differs from C++)

#
# Represents 2D vector (x,y)
# Defines dot product to get cos of angle between 2 vectors
#
class Vec2d:
    def __init__(self,x,y):
        self.x = x
        self.y = y
    def dot(self,v):
        return self.x*v.x + self.y*v.y

#
# Base class for clock hand, subclasses HourHand, MinuteHand
#
class Hand:
    # class constants to define minutes in an hour and hours on a clockface
    # may be overwritten by the using code via the set_params() method
    MINSPERHR = 60.0
    HRSPERCLOCK = 12.0

    # sets the hour/minute parameters
    @classmethod
    def set_params(cls,mph,hpc):
        cls.MINSPERHR = mph
        cls.HRSPERCLOCK = hpc

    # constructor setting time (hour or minute determined by subclass)
    def __init__(self,time):
        self.time = time

    # sets an angle with respect to the vertical
    # sets a unit vector for that angle
    def set_vector(self,angle):
        self.angle = angle
        self.vector = Vec2d( math.cos(angle), math.sin(angle))

    # calculates the difference in the angle between two hands
    # allowing/correcting for reentrant and negative angles
    def diff_angle(self, hand):
        diff = math.fabs( self.angle - hand.angle )
        if diff > twopi/2.0: diff = twopi - diff
        return diff

# Hand subclass for hour hand
class HourHand(Hand):
    # constructor setting time/hour in continuous increment by minutes
    # sets the vertical angle and its vector
    def __init__(self,hour,min):
        super().__init__(hour+min/Hand.MINSPERHR)
        super().set_vector( twopi*self.time/Hand.HRSPERCLOCK )

# Hand subclass for minute hand
class MinuteHand(Hand):
    # constructor setting time from minutes
    # and corresponding angle and vector
    def __init__(self,min):
        super().__init__(min)
        super().set_vector( twopi*self.time/Hand.MINSPERHR )

# main program asking for hour/minute input
# and printing the angle based on dot product between vectors
# and also angle based on difference between angle of handss
def main():
    # set parameters for clock
    Hand.set_params(60.0, 12.0)
    while True:
        hour = float( input("Enter hour value (<0 quits): ") )
        if hour<0: break
        min  = float( input("Enter minute value: ") )
        hhand = HourHand(hour,min)
        mhand = MinuteHand(min)
        cosangle = hhand.vector.dot( mhand.vector)
        angle = (360.0/twopi)*math.acos(cosangle)
        print(F"(dotproduct method) angle is {angle}")
        angle = (360.0/twopi)*hhand.diff_angle(mhand)
        print(F"(angledifference method) angle is {angle}")

if __name__ == '__main__':
    main()